### PR TITLE
Refactor NotJoined

### PR DESCRIPTION
**Pull request by vdimir**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by vdimir** [comment]
If CI will be ok, I'll merge it to https://github.com/ClickHouse/ClickHouse/pull/27217

**Comment by Avogar** [comment]
@vdimir can you look at Fast test failures?

**Comment by vdimir** [comment]
> @vdimir can you look at Fast test failures?

@Avogar seems it's fixed

File: src/Interpreters/HashJoin.cpp [hunk 1]
```diff
@@ -194,6 +193,13 @@ HashJoin::HashJoin(std::shared_ptr<TableJoin> table_join_, const Block & right_s
 
     required_right_keys = table_join->getRequiredRightKeys(right_table_keys, required_right_keys_sources);
 
+    LOG_DEBUG(log, "Right keys: [{}] (required: [{}]), left keys: [{}]",
+              fmt::join(key_names_right, ", "),
+              fmt::join(required_right_keys.getNames(), ", "),
+              fmt::join(table_join->keyNamesLeft(), ", "));
+
+    LOG_DEBUG(log, "Columns to add: [{}]", sample_block_with_columns_to_add.dumpStructure());
+
```
Comment (Avogar) [inline src/Interpreters/HashJoin.cpp:202]: Did you forget to remove it?

Comment (vdimir) [reply src/Interpreters/HashJoin.cpp:202]: I suppose we can leave it, it already prints 'right_sample_block' and sometimes it's good to see on what keys join performed and what columns will be added to left table.

Comment (Avogar) [reply src/Interpreters/HashJoin.cpp:202]: Ok, let's leave it

File: src/Interpreters/join_common.cpp [hunk 2]
```diff
@@ -484,49 +492,25 @@ void splitAdditionalColumns(const Names & key_names, const Block & sample_block,
...
-            if (!required_right_keys.has(right_key_name))
-                left_to_right_key_remap[left_key_pos] = right_key_pos;
-        }
-    }
-
-    /// result_sample_block: left_sample_block + left expressions, right not key columns, required right keys
-    size_t left_columns_count = result_sample_block.columns() -
-        sample_block_with_columns_to_add.columns() - required_right_keys.columns();
+    LOG_DEBUG(&Poco::Logger::get("NotJoinedBlocks"), "saved_block_sample {}",saved_block_sample.dumpStructure());
+    LOG_DEBUG(&Poco::Logger::get("NotJoinedBlocks"), "result_sample_block {}",result_sample_block.dumpStructure());
```
Comment (Avogar) [inline src/Interpreters/join_common.cpp:504]: What about these logs? It seems to me they are not supposed to remain in master :)

Comment (vdimir) [reply src/Interpreters/join_common.cpp]: Removed

**Review by Avogar — APPROVED** [review]

**Pull request merged**